### DISCUSSION TITLE
Compilation to disk IS going to happen on .NET 9 (maybe)

### DISCUSSION THREAD

**Discussion by leppie** [leppie/IronScheme#99] [Announcements]
Based on discussions, there are valid reasons why MS will not support AssemblyBuilder.Save . I do understand their unwillingness to support this, even though it is not a cross platform issue as I believed earlier. Versioning and subsequently compatibility is a huge problem with persisted assemblies created by Reflection.Emit.  When using Reflection.Emit at runtime, you cannot run into these problems as your code will fail during compilation and not runtime.

With IronScheme, there is already pretty extreme versioning requirements for persisted assemblies which avoids this problem. In fact even if nothing changes, a new bootstrap build (which happens with every release) will have a new seed for symbols, thus invalidating existing assemblies. These are safeguards I implemented originally to prevent exactly what MS is trying to prevent now. 

This validation would happen on an incremental basis so if an intermediary library changes, anything subsequently relying on that dependency would be recompiled from code instead of loading from a persisted assembly. 

That said, the only side effect having to load from code is additional compilation time (which is pretty fast already compared to other Scheme implementations).  This only has a huge time impact on libraries with macros. But once compiled and loaded, it will run just as fast. Everything else will work as intended.

Due to IronScheme using Reflection.Emit for both runtime and persisted compilation scenarios, this cannot change without changing to another IL emitter platform (ie Cecil, Roslyn, F#). I would love if someone eventually does this.

But for me, this would take quite a lot of effort that at my age of 46, I am not willing to take. IronScheme has been going for almost 14 years. First commit was Oct 11, 2007 ( https://github.com/IronScheme/IronScheme/commit/2db5223cf003bec9417662bf481378347bab77e0 ) . I can add anything you want, just not this. 

PS: There is also a known issue loading precompiled libraries (compiled from .NET desktop) in .NET Core, which is basically impossible to debug as I cannot get access to the IL that is wrong. If that can be fixed, at least we could load persisted assemblies from .NET Core solving the worst side effect.

**Comment by bcalco** [comment]
I could use a primer on designing applications and configuring projects with the different .NET targets in mind, given the compilation feature availability issue. 

Currently, I'm creating Windows desktop applications on .NET using the 4.7 target, so I'd like to know how to compile and use precompiled libraries properly, as that is available on this target, right?

I do have plans to port some utilities, mostly server side, to .NET Core on Linux and Mac., where the feature will be unavailable So understanding what needs to be different about them functionally and in terms of setting the projects up is also important to me.

Longer term, I'd like to consider becoming a maintainer and maybe evolving the compilation feature using F#, but that is not yet a plate I can start spinning, till I get the products I'm working on done first. But I want to be able to contribute in smaller ways for starters.

**Comment by leppie** [comment]
Or it might :D 

https://github.com/dotnet/runtime/issues/15704#issuecomment-975798248

**Reply by bcalco** [reply]
I have been a busy bee on other things and just saw this. At least there is hope!

**Comment by leppie** [comment]
Support is  finally coming in .NET 9.

Follow  along  here:   https://github.com/IronScheme/IronScheme/issues/158
